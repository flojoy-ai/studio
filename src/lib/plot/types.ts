export interface Drawable {
  draw: () => void;
}
