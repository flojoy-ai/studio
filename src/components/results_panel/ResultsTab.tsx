import React, { memo, useEffect, useMemo, useState } from 'react';
import ReactFlow, { ConnectionLineType, EdgeTypesType, NodeTypesType, OnLoadFunc, OnLoadParams, ReactFlowProvider } from 'react-flow-renderer';
// import Plot from 'react-plotly.js';
import { useWindowSize } from 'react-use';
import { useFlowChartState } from '../../hooks/useFlowChartState';
import CustomEdge from '../flow_chart_panel/CustomEdge';
// import styledPlotLayout from './../defaultPlotLayout';
import CustomResultNode from './CustomResultNode';
import { nodePosition } from './NODE_POSITION';

import './Results.css';

const edgeTypes: EdgeTypesType = { default: CustomEdge as any };
const nodeTypes: NodeTypesType = { default: CustomResultNode as any };

const ResultsTab = ({ results, theme }) => {
  const {width:windowWidth}= useWindowSize()
  const {rfInstance}= useFlowChartState()
  const [resultElements, setResultElements] = useState<any[]>([])
  const nodeResults = useMemo(()=> ('io' in results) 
    ? JSON.parse(results.io).reverse()
    : [],[results])

  // const styledLayout = styledPlotLayout(theme);
  const ReactFlowProviderAny: any = ReactFlowProvider;
  const onLoad: OnLoadFunc = (rfIns: OnLoadParams) => {
    rfIns.fitView();
    const flowSize = 1271;
    const xPosition =windowWidth > flowSize ? (windowWidth - flowSize) / 2 : 0;
    rfIns.setTransform({
      x: xPosition,
      y: 52,
      zoom: 0.7
    })
    // setResultElements(rfIns.toObject())
  };
useEffect(()=>{
  if(nodeResults && nodeResults.length > 0 && rfInstance){
    setResultElements(
      rfInstance?.elements.map(elem=> ({
        ...elem,
        position: nodePosition[elem?.data?.func],
        data:{
          ...elem.data,
          ...(!('source' in elem) && { resultData: nodeResults?.find(result=> result.id === elem.id)?.result})
        }
       
      }))
    )
  }
},[nodeResults, rfInstance])
  return (
    <ReactFlowProviderAny>
    <div style={{ height: `99vh` }}>
      <ReactFlow
        elements={resultElements}
        edgeTypes={edgeTypes}
        nodeTypes={nodeTypes}
        connectionLineType={ConnectionLineType.Step}
        onLoad={onLoad}
      ></ReactFlow>
    </div>
    </ReactFlowProviderAny>
    // <div className='App-results-panel'>
    //   <p>
    //     {nodeResults.length === 0 ? 'No run results yet.' : ''}
    //   </p>     
    //   {nodeResults.map((nd, k) => (
    //     <details key={k} open={ k===0 ? true : false }>
    //       <summary>{nd.cmd}</summary>
    //       {nd.result !== null ?
    //           <Plot
    //             data = {'data' in nd.result 
    //               ? nd.result.data 
    //               : [{'x': nd.result['x0'], 'y': nd.result['y0'] }]}
    //             layout = {'layout' in nd.result 
    //               ? Object.assign({}, nd.result.layout, styledLayout)
    //               : Object.assign({}, {title: `${nd.cmd}`}, styledLayout)}
    //             useResizeHandler />
    //       : `⚠️ The server returned null for the ${nd.cmd} node`}
    //     </details>
    //   ))}
    // </div>
  );
};

export default memo(ResultsTab);