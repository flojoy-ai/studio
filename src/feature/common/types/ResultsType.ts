import { OverridePlotData } from "@/types";
import { Layout } from "plotly.js";

export type ResultIO = {
  cmd: string;
  id: string;
  result: Result;
};
export type NodeResult = {
  cmd: string;
  id: string;
  result: Result;
};

export type Result = {
  plotly_fig?: {
    data: OverridePlotData;
    layout: Partial<Layout> | undefined;
  };
  text_blob?: string;
  data?: ScalarData | OrderedPairData;
};

export type ScalarData = {
  c: number;
};

export type OrderedPairData = {
  x: number[];
  y: number[];
};
