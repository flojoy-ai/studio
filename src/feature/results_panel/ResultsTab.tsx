import { memo, useEffect, useMemo, useState } from "react";
import ReactFlow, {
  ConnectionLineType,
  EdgeTypesType,
  NodeTypesType,
  OnLoadFunc,
  OnLoadParams,
  ReactFlowProvider,
} from "react-flow-renderer";
import { useWindowSize } from "react-use";
import { useFlowChartState } from "../../hooks/useFlowChartState";
import CustomEdge from "../flow_chart_panel/CustomEdge";
import CustomResultNode from "./CustomResultNode";
import { nodePosition } from "./NODE_POSITION";

import "./Results.css";

const edgeTypes: EdgeTypesType = { default: CustomEdge as any };
const nodeTypes: NodeTypesType = { default: CustomResultNode as any };

const ResultsTab = ({ results, theme }) => {
  const { width: windowWidth } = useWindowSize();
  const { rfInstance } = useFlowChartState();
  const [resultElements, setResultElements] = useState<any[]>([]);
  const nodeResults = useMemo(
    () => ("io" in results ? JSON.parse(results.io).reverse() : []),
    [results]
  );

  const ReactFlowProviderAny: any = ReactFlowProvider;
  const onLoad: OnLoadFunc = (rfIns: OnLoadParams) => {
    rfIns.fitView();
    const flowSize = 1271;
    const xPosition = windowWidth > flowSize ? (windowWidth - flowSize) / 2 : 0;
    rfIns.setTransform({
      x: xPosition,
      y: 52,
      zoom: 0.7,
    });
  };
  useEffect(() => {
    if (nodeResults && nodeResults.length > 0 && rfInstance) {
      setResultElements(
        rfInstance?.elements.map((elem) => ({
          ...elem,
          position: nodePosition[elem?.data?.func],
          data: {
            ...elem.data,
            ...(!("source" in elem) && {
              resultData: nodeResults?.find((result) => result.id === elem.id)
                ?.result,
            }),
          },
        }))
      );
    }
  }, [nodeResults, rfInstance]);
  return (
    <ReactFlowProviderAny>
      <div style={{ height: `99vh` }} data-testid="results-flow">
        <ReactFlow
          elements={resultElements}
          edgeTypes={edgeTypes}
          nodeTypes={nodeTypes}
          connectionLineType={ConnectionLineType.Step}
          onLoad={onLoad}
        ></ReactFlow>
      </div>
    </ReactFlowProviderAny>
  );
};

export default memo(ResultsTab);
