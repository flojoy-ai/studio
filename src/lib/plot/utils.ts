export function clamp(x: number, lo: number, hi: number) {
  return Math.min(Math.max(x, lo), hi);
}
