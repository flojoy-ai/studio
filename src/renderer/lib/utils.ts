import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";
import { z } from "zod";
import * as E from "fp-ts/Either";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function tryParse<Z extends z.ZodTypeAny>(z: Z) {
  return (val: unknown) => {
    const res = z.safeParse(val);
    if (res.success) {
      return E.right(res.data as z.infer<Z>);
    }
    return E.left(res.error);
  };
}
