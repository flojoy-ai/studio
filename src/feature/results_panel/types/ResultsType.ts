import { ElementsData } from "@src/feature/flow_chart_panel/types/CustomNodeProps";
import { Data, Layout } from "plotly.js";

export type ResultIO = {
  cmd: string;
  id: string;
  result: {
    default_fig: {
      data: Data[],
      layout: Partial<Layout>
    },
    data: {
      type: string;
      [key:string]: any;
    }
  };
};
export interface ResultsType {
  io?: ResultIO[];
}
export type ResultNodeData = ElementsData & {
  resultData?: ResultIO["result"];
};
