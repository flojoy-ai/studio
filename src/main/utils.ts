import * as http from "http";
import * as fs from "fs";
import { isIP } from "net";
import { execCommand } from "./executor";
import { Command } from "./command";
import { app, dialog } from "electron";
import { join } from "path";
import { killCaptain } from "./python";
import log from "electron-log";
import { ChildProcess } from "node:child_process";
import { Err, Result, Ok } from "../types/result";

export const isPortFree = (port: number) =>
  new Promise((resolve) => {
    const server = http
      .createServer()
      .listen(port, "127.0.0.1", () => {
        server.close();
        resolve(true);
      })
      .on("error", () => {
        resolve(false);
      });
  });

export const killProcess = async (port: number) => {
  await execCommand(
    new Command({
      darwin: `kill -9 $(lsof -t -i :${port})`,
      linux: `kill -9 $(lsof -t -i :${port})`,
      win32: `FOR /F "tokens=5" %i IN ('netstat -aon ^| find "${port}"') DO Taskkill /F /PID %i`,
    }),
  );
};

export const ping = async (addr: string) => {
  if (isIP(addr) === 0) {
    throw new Error(`Invalid IP address: ${addr}`);
  }

  return await execCommand(
    new Command({
      win32: `ping -n 1 ${addr}`,
      darwin: `ping -c 1 ${addr}`,
      linux: `ping -c 1 ${addr}`,
    }),
    { quiet: true },
  );
};

export const netstat = async () => {
  return await execCommand(
    new Command({
      win32: "netstat",
      darwin: "netstat",
      linux: "netstat",
    }),
    { quiet: true },
  );
};

export const ifconfig = async () => {
  return await execCommand(
    new Command({
      win32: "ipconfig /all",
      darwin: "ifconfig",
      linux: "ifconfig",
    }),
    { quiet: true },
  );
};

export const writeFileSync = (_, filePath: string, text: string): void => {
  fs.writeFileSync(filePath, text);
};

export const pickDirectory = async (): Promise<string> => {
  const handler = await dialog.showOpenDialog({
    properties: ["openDirectory"],
  });
  return handler.canceled ? "" : handler.filePaths[0];
};

export const getCustomBlocksDir = async () => {
  const filePath = join(app.getPath("home"), ".flojoy/custom_blocks_path.txt");

  if (!fs.existsSync(filePath)) {
    return undefined;
  }

  const blocksPath = fs
    .readFileSync(filePath, { encoding: "utf-8" })
    .toString();

  if (!fs.existsSync(blocksPath)) {
    return undefined;
  }

  return blocksPath;
};

export const cacheCustomBlocksDir = (_, dirPath: string) => {
  const flojoyDir = join(app.getPath("home"), ".flojoy");
  if (!fs.existsSync(flojoyDir)) {
    fs.mkdirSync(flojoyDir);
  }
  const cacheFilePath = join(flojoyDir, "custom_blocks_path.txt");
  fs.writeFileSync(cacheFilePath, dirPath);
};

export const openTestPicker = (): Promise<
  { filePath: string; fileContent: string } | undefined
> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: ["json", "py"],
            name: "Test",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const openFilePicker = (
  _,
  allowedExtensions: string[] = ["json"],
): Promise<{ filePath: string; fileContent: string } | undefined> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: allowedExtensions,
            name: "allowed extensions",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
        resolve(undefined);
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const cleanup = async () => {
  const captainProcess = global.captainProcess as ChildProcess;
  log.info(
    "Cleanup function invoked, is captain running? ",
    !(captainProcess?.killed ?? true),
  );
  if (captainProcess && captainProcess.exitCode === null) {
    const success = killCaptain();
    if (success) {
      global.captainProcess = null;
      log.info("Successfully terminated captain :)");
    } else {
      log.error("Something went wrong when terminating captain!");
    }
  }
};

export const loadFileFromFullPath = async (
  filePath: string,
): Promise<string> => {
  return fs.readFileSync(filePath, { encoding: "utf-8" }).toString();
};

export const saveFileToFullPath = async (
  filePath: string,
  content: string,
): Promise<Result<void>> => {
  try {
    fs.writeFileSync(filePath, content);
    return Ok(undefined);
  } catch (error) {
    return Err(error as Error);
  }
};
