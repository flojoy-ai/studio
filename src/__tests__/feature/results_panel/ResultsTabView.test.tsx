import { screen } from "@testing-library/react";
import ResultsTab from "@src/feature/results_panel/ResultsTabView";
import { ResultsType } from "@src/feature/results_panel/types/ResultsType";
import { renderWithTheme } from "@src/__tests__/__utils__/utils";

jest.mock("@src/hooks/useFlowChartState");
jest.mock("@src/hooks/useControlsState");

jest.mock("@src/feature/results_panel/ResultsTabEffects", () => ({
  useResultsTabEffects: jest.fn(),
}));

jest.mock("@src/feature/results_panel/ResultsTabState", () => ({
  useResultsTabState: jest.fn(() => ({
    setResultNodes: jest.fn(),
    resultNodes: [] as Node[],
    nodes: [] as Node[],
  })),
}));

jest.mock("@src/feature/results_panel/views/CustomResultNode", () => {
  return <div></div>;
});
jest.mock("@src/feature/flow_chart_panel/views/CustomEdge", () => {
  return <div></div>;
});
jest.mock("react-use", () => ({
  useWindowSize: jest.fn(() => ({ width: 1024, height: 768 })),
}));

jest.mock("@src/hooks/useSocket");

describe("ResultsTab component", () => {
  const results: ResultsType = {
    io: [],
  };

  it("renders the ResultsTab component", () => {
    renderWithTheme(<ResultsTab />);

    const reactFlowComponent = screen.getByTestId("results-flow");
    expect(reactFlowComponent).toBeInTheDocument();
    expect(screen.getByTestId("react-flow")).toBeInTheDocument();

    expect(reactFlowComponent).toHaveAttribute(
      "style",
      "height: calc(100vh - 110px);"
    );
    expect(reactFlowComponent).toMatchSnapshot();
  });
});
