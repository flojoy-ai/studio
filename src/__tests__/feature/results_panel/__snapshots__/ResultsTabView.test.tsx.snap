// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsTab component renders the ResultsTab component 1`] = `
<div
  data-testid="results-flow"
  style="height: calc(100vh - 110px);"
>
  <div
    data-testid="react-flow"
  />
</div>
`;
